numpy/lib/nanfunctions.py

```diff
@@ -1235,3 +1236,114 @@ def nanstd(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
+    of :math:`x_i`.
+
+    Parameters
+    ----------
+    m : array_like
+        A 1-D or 2-D array containing multiple variables and observations.
+        Each row of `m` represents a variable, and each column a single
+        observation of all those variables. Also see `rowvar` below.
+    y : array_like, optional
+        An additional set of variables and observations. `y` has the same
+        form as that of `m`.
+    rowvar : int, optional
+        If `rowvar` is non-zero (default), then each row represents a
+        variable, with observations in the columns. Otherwise, the relationship
+        is transposed: each column represents a variable, while the rows
+        contain observations.
+    bias : int, optional
+        Default normalization is by ``(N - 1)``, where ``N`` is the number of
+        observations given (unbiased estimate). If `bias` is 1, then
+        normalization is by ``N``. These values can be overridden by using
+        the keyword ``ddof`` in numpy versions >= 1.5.
+    ddof : int, optional
+        .. versionadded:: 1.5
+        If not ``None`` normalization is by ``(N - ddof)``, where ``N`` is
+        the number of observations; this overrides the value implied by
+        ``bias``. The default value is ``None``.
+
+    Returns
+    -------
+    out : ndarray
+        The covariance matrix of the variables.
+
+    See Also
+    --------
+    corrcoef : Normalized covariance matrix
+
+    Examples
+    --------
+    Consider two variables, :math:`x_0` and :math:`x_1`, which
+    correlate perfectly, but in opposite directions:
+
+    >>> x = np.array([[0, 2], [1, 1], [np.nan, 1], [2, 0]]).T
+    >>> x
+    array([[0, 1, nan, 2],
+           [2, 1,   1, 0]])
+
+    Note how :math:`x_0` increases while :math:`x_1` decreases. The covariance
+    matrix shows this clearly:
+
+    >>> np.nancov(x)
+    array([[ 1., -1.],
+           [-1.,  1.]])
+
+    Note that element :math:`C_{0,1}`, which shows the correlation between
+    :math:`x_0` and :math:`x_1`, is negative.
+
+    Further, note how `x` and `y` are combined:
+
+    >>> x = [-2.1, -1,  np.nan, 4.3]
+    >>> y = [3,  1.1,  1, 0.12]
+    >>> X = np.vstack((x,y))
+    >>> print np.nancov(X)
+    [[ 11.71        -4.286     ]
+     [ -4.286        2.14413333]]
+    >>> print np.nancov(x, y)
+    [[ 11.71        -4.286     ]
+     [ -4.286        2.14413333]]
+    >>> print np.nancov(x)
+    11.71
+
+    """
+    m = np.asarray(m)
+    if y is None:
+        dtype = np.result_type(m, np.float64)
+    else:
+        y = np.asarray(y)
+        dtype = np.result_type(m, y, np.float64)
+    X = np.array(m, ndmin=2, dtype=dtype)
+
+    if X.shape[0] == 1:
+        rowvar = 1
+    if rowvar:
+        axis = 0
+    else:
+        axis = 1
+
+    if y is not None:
+        y = np.array(y, copy=False, ndmin=2, dtype=dtype)
+        X = np.concatenate((X, y), axis)
+
+    # Remove observations with nans from the set
+    nan_observations = np.isnan(X.sum(axis=axis))
```

Comment:
There isn't are multiple reasonable definitions of the covariance matrix in
the presence of missing values. E.g. including all pairs that contain no
nan gives you the best estimate of individual entries, but the matrix may
then fail to be positive definite. R provides multiple methods, see the
use= argument:
https://stat.ethz.ch/R-manual/R-patched/library/stats/html/cor.html
On Mar 19, 2015 7:15 PM, "dfreese" <notifications@github.com> wrote:

> In numpy/lib/nanfunctions.py
> <https://github.com/numpy/numpy/pull/5698#discussion_r26814593>:
>
> > +        dtype = np.result_type(m, y, np.float64)
> > +    X = np.array(m, ndmin=2, dtype=dtype)
> > +
> > +    if X.shape[0] == 1:
> > +        rowvar = 1
> > +    if rowvar:
> > +        axis = 0
> > +    else:
> > +        axis = 1
> > +
> > +    if y is not None:
> > +        y = np.array(y, copy=False, ndmin=2, dtype=dtype)
> > +        X = np.concatenate((X, y), axis)
> > +
> > +    # Remove observations with nans from the set
> > +    nan_observations = np.isnan(X.sum(axis=axis))
>
> Now that you point that out, no, it's not correct. Looking at the problem,
> though, I don't see a great way to accomplish this except for iterating
> over every combination of the rows, since each combination could, in
> theory, have a unique set of NaNs. However, that doesn't seem terribly
> elegant.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5698/files#r26814593>.
>